app/models/manageiq/providers/openstack/infra_manager/host.rb

```diff
@@ -35,48 +40,84 @@ def get_parent_keypair(type = nil)
   def authentication_best_fit(type = nil)
     # First check for Host level credentials with filled private key, this way we can override auth credentials per
     # host. Otherwise host level auth will hold only state of auth for the given host. We allow only auth with private
-    # key for OpenstackInfra hosts
-    auth = authentication_type(type)
-    return auth if auth && auth.auth_key
+    # key and password for OpenstackInfra hosts
+    auth = if !authentication_type(type).try(:password).blank? || !authentication_type(type).try(:auth_key).blank?
+             authentication_type(type)
+           elsif !authentication_type(:ssh_keypair).try(:auth_key).blank?
+             authentication_type(:ssh_keypair)
+           elsif !authentication_type(:default).try(:password).blank?
+             authentication_type(:default)
+           end
+
+    return auth if auth && (auth.password || auth.auth_key)
```

Comment:
This block has a lot of calls to `authentication_type` and checks against `password` and `auth_key` fields.

Not really sure if this is better, but it seems easier to read to me... 

thoughts?

```ruby
# override from authentication_mixin
def authentication_type(type)
  auth = super
  validate_authentication_by_type(auth)
end

def validate_authentication_by_type(auth)
  return unless auth
  case auth.authentication_type
  when :ssh_keypair
    auth unless auth.auth_key.blank?
  when :default
    auth unless auth.password.blank?
  else # no specific rule for this auth type, check both fields
    auth unless (auth.auth_key.blank? && auth.password.blank?)
  end
end

def authentication_best_fit(type = nil)
  authentication_type(type) || get_parent_keypair(:ssh_keypair) 
end
```